### PR TITLE
[TG-1523] fix removed required virtual calls

### PR DESCRIPTION
**Pull request by mgudemann**
first version of approach to solve the problem of parent classes implementing an interface

### CODE REVIEW

**Comment by mgudemann** [comment]
Current state is as follows, the dispatch table is analyzed for entries that have a empty `symbol_exprt` as their `symbol_expr` field.
For those cases the `get_parent_functions` analogon to `get_child_functions` is called, if a result for an implementing class is found, these are replaced in the dispatch table.

For `ArrayList` this results in the following:
```
problematic dispatch table
java::java.util.SubList$1:symbol java::java.util.ArrayList$SubList$1:symbol java::java.util.ArrayList$ListItr:symbol java::java.util.AbstractList$ListItr:symbol 
nothing found for java::java.util.SubList$1
nothing found for java::java.util.ArrayList$SubList$1
changed dispatch table
java::java.util.SubList$1:symbol java::java.util.ArrayList$SubList$1:symbol java::java.util.ArrayList$ListItr:symbol
  * type: code
      * return_type: c_bool
          * width: 8
      * access: public
      * parameters: 
        0: parameter
            * type: pointer
                * width: 64
                * #reference: 1
                0: symbol
                    * identifier: java::java.util.ArrayList$Itr
            * #base_name: this
            * #identifier: java::java.util.ArrayList$Itr.hasNext:()Z::this
            * #this: 1
  * identifier: java::java.util.ArrayList$Itr.hasNext:()Z
  * #class: java::java.util.ArrayList$Itr java::java.util.AbstractList$ListItr:symbol
  * type: code
      * return_type: c_bool
          * width: 8
      * access: public
      * parameters: 
        0: parameter
            * type: pointer
                * width: 64
                * #reference: 1
                0: symbol
                    * identifier: java::java.util.AbstractList$Itr
            * #base_name: this
            * #identifier: java::java.util.AbstractList$Itr.hasNext:()Z::this
            * #this: 1
  * identifier: java::java.util.AbstractList$Itr.hasNext:()Z
  * #class: java::java.util.AbstractList$Itr 
```
that is for none of the candidate classes there's an implementing function. Calling `get_parent_functions` return a valid entry for `java.util.ArrayList$ListItr`, namely its parent class `java.util.ArrayList$Itr`, but nothing for the others.

In the GOTO this results in the following code
```
        // 804 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        ASSUME !(e2 == null) // Throw null
        // 805 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        IF "java::java.util.ArrayList$ListItr" == e2->@java.lang.Object.@class_identifier THEN GOTO 15
        // 806 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        IF "java::java.util.ArrayList$SubList$1" == e2->@java.lang.Object.@class_identifier THEN GOTO 16
        // 807 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        IF "java::java.util.SubList$1" == e2->@java.lang.Object.@class_identifier THEN GOTO 16
        // 808 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        (struct java.util.AbstractList$Itr *)e2 . java.util.AbstractList$Itr.hasNext:()Z();
        // 809 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        return_tmp3 = java.util.AbstractList$Itr.hasNext:()Z#return_value;
        // 810 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        dead java.util.AbstractList$Itr.hasNext:()Z#return_value;
        // 811 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        GOTO 17
        // 812 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
    15: (struct java.util.ArrayList$Itr *)e2 . java.util.ArrayList$Itr.hasNext:()Z();
        // 813 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        return_tmp3 = java.util.ArrayList$Itr.hasNext:()Z#return_value;
        // 814 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        dead java.util.ArrayList$Itr.hasNext:()Z#return_value;
        // 815 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
        GOTO 17
        // 816 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 21
    16: ASSUME false // cannot find calls for java::java.util.ListIterator.hasNext:()Z dispatching java::java.util.ArrayList$SubList$1
        // 817 file java/util/AbstractList.java line 544 function java::java.util.AbstractList.equals:(Ljava/lang/Object;)Z bytecode-index 22
```
That is, for the class_id in question, the parent classes' implementation is called and the non-existence of another implementation is marked in the comment of the generated `ASSERT / ASSUME false`.

@smowton as you certainly have more insight into how this should work, is this a viable solution?

Unfortunately I failed to construct a simple regression test for this. Building something where a child class implements the interface and its parent class actually implements the interface's method, CBMC created a non-det version of the function for the interface.

**Comment by tautschnig** [comment]
@mgudemann Would you be able to provide a complicated regression test (you said you failed to build a simple one)? I think it would be great if anything vaguely resembling a bugfix also came with a regression test.

File: src/goto-programs/class_hierarchy.cpp [hunk 1]
```diff
@@ -99,3 +104,8 @@ void class_hierarchyt::output(std::ostream &out) const
           << ch << '\n';
   }
 }
+
+bool class_hierarchyt::is_abstract_class(const irep_idt &c) const
+{
+  return abstract_classes.count(c) > 0;
```
Comment (tautschnig) [inline src/goto-programs/class_hierarchy.cpp]: I understand that using `count` requires less typing, but then it does require looking up the container type to figure out what's going on. Would you mind using `abstract_classes.find(c) != abstract_classes.end()` instead?

File: src/goto-programs/class_hierarchy.cpp [hunk 2]
```diff
@@ -99,3 +104,8 @@ void class_hierarchyt::output(std::ostream &out) const
...
+}
```
Comment (tautschnig) [inline src/goto-programs/class_hierarchy.cpp]: I might not have a sufficient understanding of this code, but it seems the approach taken here has potential for problems: this function may return different answers depending on _when_ it is called.

I'd strongly recommend doing lookups in the `symbol_table` instead, thus avoiding the need for `abstract_classes` altogether.

File: src/goto-programs/class_hierarchy.h [hunk 3]
```diff
@@ -52,10 +54,12 @@ class class_hierarchyt
   }
 
   void output(std::ostream &) const;
+  bool is_abstract_class(const irep_idt &) const;
 
 protected:
   void get_children_trans_rec(const irep_idt &, idst &) const;
   void get_parents_trans_rec(const irep_idt &, idst &) const;
+  bool has_abstract_bases(const irep_idt &c) const;
```
Comment (tautschnig) [inline src/goto-programs/class_hierarchy.h]: Where is this implemented?

File: src/goto-programs/class_hierarchy.cpp (hunk 2 above)
Comment (mgudemann) [reply src/goto-programs/class_hierarchy.cpp]: @tautschnig definitely right with both comments

File: src/goto-programs/class_hierarchy.h (hunk 3 above)
Comment (mgudemann) [reply src/goto-programs/class_hierarchy.h]: will go away, isn't used currently

**Comment by mgudemann** [comment]
@tautschnig the known case where this problem appears uses proprietary extensions unfortunately, the basic structure is the one outlined above, a class that implements an interface where the necessary methods actually come from a base class.

File: src/goto-programs/remove_virtual_functions.cpp [hunk 4]
```diff
@@ -322,13 +384,99 @@ void remove_virtual_functionst::get_functions(
...
   std::set<irep_idt> visited;
-  get_child_functions_rec(
+  get_child_functions(
     class_id,
     root_function.symbol_expr,
     component_name,
     functions,
     visited);
 
+  const auto find_empty = [&](const dispatch_table_entryt &entry) {
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: Just capture `this`?

File: src/goto-programs/remove_virtual_functions.cpp [hunk 5]
```diff
@@ -322,13 +384,99 @@ void remove_virtual_functionst::get_functions(
...
+    return entry.symbol_expr == symbol_exprt();
+  };
+  bool empty_found =
+    std::any_of(functions.begin(), functions.end(), find_empty);
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: Name the lambda `is_empty`

File: src/goto-programs/remove_virtual_functions.cpp [hunk 6]
```diff
@@ -47,12 +48,20 @@ class remove_virtual_functionst
...
-  void get_child_functions_rec(
+  void get_child_functions(
     const irep_idt &,
     const symbol_exprt &,
     const irep_idt &,
     dispatch_table_entriest &,
-    std::set<irep_idt> &visited) const;
+    std::set<irep_idt> &visited,
+    bool recurse=true) const;
+  void get_parent_functions(
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: This reads like a duplicate of `resolve_concrete_function_callt` (see https://github.com/diffblue/cbmc/blob/develop/src/goto-programs/remove_virtual_functions.cpp#L300 for example)

File: src/goto-programs/remove_virtual_functions.cpp [hunk 7]
```diff
@@ -267,6 +281,49 @@ void remove_virtual_functionst::get_child_functions_rec(
...
+    }
+    else
+    {
+      function.symbol_expr=last_method_defn;
+    }
+    functions.push_back(function);
+
+    if(recurse)
+    {
+      get_child_functions(
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: Why would you ever want to not recurse over child types?

File: src/goto-programs/remove_virtual_functions.cpp [hunk 8]
```diff
@@ -280,12 +337,15 @@ void remove_virtual_functionst::get_child_functions_rec(
...
 
-    get_child_functions_rec(
-      child,
-      function.symbol_expr,
-      component_name,
-      functions,
-      visited);
+    if(recurse)
+    {
+      get_parent_functions(
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: As above, get_parent_functions is probably a duplicate of the existing resolve-concrete method.

File: src/goto-programs/remove_virtual_functions.cpp [hunk 9]
```diff
@@ -322,13 +384,99 @@ void remove_virtual_functionst::get_functions(
...
+    std::any_of(functions.begin(), functions.end(), find_empty);
+
+  if(empty_found && !functions.empty())
+  {
+    std::cout << "problematic dispatch table\n";
+    for(const auto &f : functions)
+      std::cout << f.class_id << ":" << f.symbol_expr.pretty() << " ";
+    std::cout << std::endl;
+
+    // with a non-empty virtual dispatch table but only empty function
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: The condition says *any* class fails to provide a definition, but the comment says *all*

File: src/goto-programs/remove_virtual_functions.cpp [hunk 10]
```diff
@@ -322,13 +384,99 @@ void remove_virtual_functionst::get_functions(
...
+    // expressions, we could not find a class that provides the required
+    // method. Therefore we try finding it in the set of parent classes of a
+    // class in the dispatch table.
+
+    std::map<irep_idt, symbol_exprt> new_entries;
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: Why are we doing this as a fixup step, not getting it right the first time? During the walk over children I could simply ask at each node, if there is no definition at this child, use resolve-concrete to search for an inherited definition

File: src/goto-programs/remove_virtual_functions.cpp [hunk 11]
```diff
@@ -322,13 +384,99 @@ void remove_virtual_functionst::get_functions(
...
+            std::find_if_not(entries.begin(), entries.end(), find_empty);
+          if(it != entries.end())
+            new_entries[f.class_id] = it->symbol_expr;
+          else
+            std::cout << "nothing found for " << f.class_id << std::endl;
+      //   }
+      // }
+    }
+
+    // collect sets of children classes for each class in the new additional
```
Comment (smowton) [inline src/goto-programs/remove_virtual_functions.cpp]: I think this is incorrect. If we have an interface `Interface` that prototypes a method `foo`, and we have types `A`, `B` and `C`, such that (writing `<:` for subtyping) `A <: C, B <: C, A <: Interface`, and `C` and `B` define concrete definitions of `foo`, then a call `((Interface)something).foo()` may target an `A` object, and thus call `C.foo`, but may *not* target a `B` since that doesn't inherit `Interface`. This loop would appear to gather `B` as a candidate callee because it is a child of `C`.

In short, it should suffice to try to resolve a concrete function call at each child node where the method is abstract, without need for this further search over *child* types.

**Review by smowton — CHANGES_REQUESTED** [review]
I think this is currently incorrect. Recommend simplifying and using existing code to check for inherited methods when a method is abstract at a particular point in the class hierarchy; beyond that no structural change should be necessary.

File: src/goto-programs/remove_virtual_functions.cpp (hunk 7 above)
Comment (mgudemann) [reply src/goto-programs/remove_virtual_functions.cpp]: I had abused this to check only direct parents, but this is now the recursive `get_parent_function` method (and probably soon the one you pointed out)

File: src/goto-programs/remove_virtual_functions.cpp (hunk 11 above)
Comment (mgudemann) [reply src/goto-programs/remove_virtual_functions.cpp]: @smowton thanks for your comments, I now use resolve-concrete-function which does this in a better way. For each child, I check for such an implementation if get_child_functions was not successful in finding one.

Comment (smowton) [reply src/goto-programs/remove_virtual_functions.cpp]: I think it should check for one *even if* get-child-functions finds one. Example:

C extends B extends A. B implements Interface. C and A provide implementations of f. Then a call to ((Interface)x).f() should consider B.f() aka A.f() as a target, even though its child C provides an implementation.

File: src/goto-programs/remove_virtual_functions.cpp [hunk 12]
```diff
@@ -52,10 +53,13 @@ class remove_virtual_functionst
     const symbol_exprt &,
     const irep_idt &,
     dispatch_table_entriest &,
-    std::set<irep_idt> &visited) const;
-  exprt get_method(
-    const irep_idt &class_id,
-    const irep_idt &component_name) const;
+    std::set<irep_idt> &visited,
+    const std::function<
```
Comment (thk123) [inline src/goto-programs/remove_virtual_functions.cpp]: Consider typedef'ing this

File: src/goto-programs/remove_virtual_functions.cpp [hunk 13]
```diff
@@ -281,14 +290,30 @@ void remove_virtual_functionst::get_child_functions_rec(
...
     }
+    if(function.symbol_expr == symbol_exprt())
+    {
+      const resolve_concrete_function_callt::concrete_function_callt
+        &resolved_call = resolve_function_call(child, component_name);
+      if(resolved_call.is_valid())
+      {
+        function.class_id = resolved_call.get_class_identifier();
+        const symbolt &called_symbol =
+          *symbol_table.lookup(resolved_call.get_virtual_method_name());
```
Comment (majakusber) [inline src/goto-programs/remove_virtual_functions.cpp]: Can you use `lookup_ref` instead?

**Review by majakusber — COMMENTED** [review]
As already suggested, it would be great to have a test for this. If not a regression test, how about a unit test? Chris mentioned few class/interface hierarchies that should be caught correctly in his comments, can they be used to see is the search finds the correct function?

**Review by thk123 — COMMENTED** [review]
Will do a re-review once existing comments are addressed.

**Comment by mgudemann** [comment]
@smowton @svorenova @tautschnig I added a regression test for this. Interestingly, when using the orignal version, there's a call to `ListIterator.hasNext` and with `--lazy-loading` the call disappears. With the PR, the correct call to `ArrayList$Itr.hasNext` is added.

**Comment by tautschnig** [comment]
Is the "Do not merge" still accurate? (The linter failure will certainly need addressing, but otherwise it sounded like this was ready for another review by @smowton.)

File: src/goto-programs/class_hierarchy.cpp [hunk 14]
```diff
@@ -31,6 +31,11 @@ void class_hierarchyt::operator()(const symbol_tablet &symbol_table)
       const struct_typet &struct_type=
         to_struct_type(it->second.type);
 
+      if(to_class_type(struct_type).is_abstract())
+      {
+        abstract_classes.insert(it->first);
```
Comment (smowton) [inline src/goto-programs/class_hierarchy.cpp]: Unused?

File: src/goto-programs/class_hierarchy.h [hunk 15]
```diff
@@ -32,6 +33,7 @@ class class_hierarchyt
 
   typedef std::map<irep_idt, entryt> class_mapt;
   class_mapt class_map;
+  std::set<irep_idt> abstract_classes;
```
Comment (smowton) [inline src/goto-programs/class_hierarchy.h]: Again, unused?

**Review by smowton — COMMENTED** [review]
Patch looks reasonable, I think the `is_abstract_class` stuff is unused though?

**Comment by mgudemann** [comment]
@tautschnig @smowton adressed your comments, linter should be happy now and the `is_abstract` stuff is removed

**Comment by tautschnig** [comment]
It seems the linter isn't yet perfectly happy, and I'm still wondering about do-not-merge?

**Comment by mgudemann** [comment]
@tautschnig we switched to clang-format and there are cases where clan-format and cpplint do not agree, e.g., in this case of where to put braces for a lambda function

clang-format says
```
  const function_call_resolvert resolve_function_call =
    [&get_virtual_call_target](
      const irep_idt &class_id, const irep_idt &component_name) {
      return get_virtual_call_target(class_id, component_name);
    };
```
while cpplint would like the opening brace `{` on a separate line. As far as I know in this case we go with clang-format. cpplint doesn't cause a real CI failure, but clang-format does now. I could add a linter-exception, but would prefer not to, in order to prevent later changes and linter problems to go unnoticed.

File: src/goto-programs/class_hierarchy.h [hunk 16]
```diff
@@ -16,6 +16,7 @@ Date: April 2016
 
 #include <iosfwd>
 #include <map>
+#include <set>
```
Comment (smowton) [inline src/goto-programs/class_hierarchy.h]: Unnecessary

**Review by smowton — APPROVED** [review]
One line of probably unnecessary change; otherwise LGTM, no need for re-review.

**Comment by tautschnig** [comment]
> cpplint doesn't cause a real CI failure, but clang-format does now.

That's news to me, and this PR seems proof of that: the "Cpp Linter" job certainly failed and is blocking.

> I could add a linter-exception, but would prefer not to, in order to prevent later changes and linter problems to go unnoticed.

As far as I recall a similar discussion has happened on other PRs; either someone needs to sit down and adjust the linter, or else (much easier) just add `// NOLINTNEXTLINE(whitespace/braces)` -- I don't think that would cause other linter problems to go unnoticed?

**Comment by mgudemann** [comment]
@tautschnig and the do not merge just went away and this could be merged once reviewers are ok

**Comment by mgudemann** [comment]
@tautschnig cpplint failures do not prevent merging, in contrast to clang-format failures which are part of Travis CI runs

**Review by majakusber — APPROVED** [review]
Looks good, thank you for adding the test.

File: src/goto-programs/remove_virtual_functions.cpp [hunk 17]
```diff
@@ -294,13 +322,22 @@ void remove_virtual_functionst::get_functions(
...
-  const resolve_concrete_function_callt::concrete_function_callt &
-    resolved_call=get_virtual_call_target(class_id, component_name);
+  const function_call_resolvert resolve_function_call =
+    [&get_virtual_call_target](
+      const irep_idt &class_id, const irep_idt &component_name) {
+      return get_virtual_call_target(class_id, component_name);
+    };
+
+  const resolve_concrete_function_callt::concrete_function_callt
+    &resolved_call = resolve_function_call(class_id, component_name);
```
Comment (thk123) [inline src/goto-programs/remove_virtual_functions.cpp]: Why not just call `get_virtual_call_target`?

File: src/goto-programs/remove_virtual_functions.cpp [hunk 18]
```diff
@@ -278,14 +290,30 @@ void remove_virtual_functionst::get_child_functions_rec(
...
+      if(resolved_call.is_valid())
+      {
+        function.class_id = resolved_call.get_class_identifier();
+        const symbolt &called_symbol =
+          symbol_table.lookup_ref(resolved_call.get_virtual_method_name());
+
+        function.symbol_expr = called_symbol.symbol_expr();
+        function.symbol_expr.set(
+          ID_C_class, resolved_call.get_class_identifier());
+      }
```
Comment (thk123) [inline src/goto-programs/remove_virtual_functions.cpp:306]: What does an invalid `resolved_call` mean? Should this be an invariant or is this reasonable behaviour? If reasonable, would be good to have a to have a test that explores this path.

Comment (mgudemann) [reply src/goto-programs/remove_virtual_functions.cpp:306]: an unresolved call should not simply remove the function call, but t least call a non-deterministic function, like it seems to be done when not using `--lazy-methods`

File: src/goto-programs/remove_virtual_functions.cpp (hunk 17 above)
Comment (mgudemann) [reply src/goto-programs/remove_virtual_functions.cpp]: done

**Review by thk123 — APPROVED** [review]

**Comment by mgudemann** [comment]
@tautschnig the difference between clang-format and cpplint isn't ideal, I agree. But then, these cases seem to be quite limited in number.

**Comment by tautschnig** [comment]
> But then, these cases seem to be quite limited in number.

I think it's any use of a lambda and the colon in class declarations/inheritance, as far as I can tell.

But then I don't know what to deduce from that comment?
